File: BinarySearch/MinimumTripTime/tripTime.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    bool check(long long trip,vector<int>& time, long long tripTime){
        long long totalTrip = 0;
        for(int i=0;i<time.size();i++){
            totalTrip += tripTime/(long long)time[i];
        }
        return trip <= totalTrip;
    }
    long long minimumTime(vector<int>& time, int totalTrips) {
        long long lo = 1;
        long long hi = (long long)*min_element(time.begin(), time.end()) * totalTrips;
        long long ans = -1;
        while(lo<=hi){
            long long mid = lo+(hi-lo)/2;
            if(check(totalTrips,time,mid)){
                ans = mid;
                hi = mid-1;
            }else lo = mid+1;
        }
        return ans;
    }
};

File: BinarySearch/MinimumTripTime/tripTime.py
from typing import List

class Solution:
    def check(self,trip,time,tripTime):
        totalTrip = 0
        for i in range(len(time)):
            totalTrip += tripTime//time[i]
        return trip <= totalTrip

    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        lo = 1
        hi = min(time) * totalTrips
        ans = -1
        while lo<=hi:
            mid = lo+(hi-lo)//2
            if self.check(totalTrips,time,mid):
                ans = mid
                hi = mid-1
            else: lo = mid+1
        return ans